src/lib.rs

```diff
@@ -472,29 +472,30 @@ pub fn create(value: &EtherValue, data: &[u8]) -> CreateResult {
 }
 
 /// Executes callDataCopy, but does not check for overflow.
-pub fn unsafe_calldata_copy(from: usize, length: usize) -> Vec<u8> {
-    let mut ret: Vec<u8> = unsafe_alloc_buffer(length);
-
+pub fn unsafe_calldata_copy(from: usize, length: usize, ret: &mut [u8]) {
     unsafe {
         native::ethereum_callDataCopy(ret.as_mut_ptr() as *const u32, from as u32, length as u32);
     }
-
-    ret
 }
 
 /// Returns a vector containing all data passed with the currently executing call.
 pub fn calldata_acquire() -> Vec<u8> {
-    unsafe_calldata_copy(0, calldata_size())
+    let length = calldata_size();
+
+    let mut ret: Vec<u8> = unsafe_alloc_buffer(length);
```

Comment:
What is the purpose of using `unsafe_alloc_buffer`? It seems like you can do this safely using `From` or `slice.to_vec()`

Reply:
Uhm, have you seen the code? :)

Reply:
that's not exactly a very effective way of pointing out the problem.

Reply:
After the change it is only used by the `_acquire` functions. By requiring them to use a slice defeats their purpose, one could just use the underlying functions.

I'm wondering if we keep the `_acquire` ones in with `std::Vec`, but the caller doesn't use it, will Rust properly optimise `std::Vec` out?

Reply:
Then again we could make it depend on `#feature(!no_std)`